### DISCUSSION TITLE
TypeError: Cannot read properties of undefined (reading 'bind') when using headers() in production

### DISCUSSION THREAD

**Discussion by MilanCiprian** [vercel/next.js#72466] [Help]
### Link to the code that reproduces this issue

https://github.com/MilanCiprian/headers_test

The error occurs when accessing headers() in production. `const requestHeaders = await headers();` from file src/app/_util/getIP.ts

### To Reproduce

1. Run `npm run build`
2. Run `npm run start`
3. Access the app on http://localhost:3000/

**File Structure:**

* src/app/page.tsx
```typescript
"use server"
import { getIP } from "./_util/getIP";

export default async function Home() {
  const ip = await getIP();
  return (
    <h1>TEST IP {ip}</h1>
  );
}
```

* src/app/_util/getIP.ts
```typescript
"use server"
import { headers } from "next/headers";

export async function getIP() {
  const requestHeaders = await headers(); // this line generate the error
  const remoteIP = requestHeaders.get("x-forwarded-for")?.split(":").pop() || "unknown";
  return remoteIP;
}
```

### Current vs. Expected behavior

It works fine in development mode (npm run dev), but after building (npm run build) and starting the app (npm run start), I encounter this error: `⨯ TypeError: Cannot read properties of undefined (reading 'bind')`

Expected display:
TEST IP 1

### Provide environment information

```bash
Operating System:
  Platform: linux
  Arch: x64
  Version: #1 SMP PREEMPT_DYNAMIC Debian 6.1.69-1 (2023-12-30)
  Available memory (MB): 15786
  Available CPU cores: 12
Binaries:
  Node: 18.19.0
  npm: 9.2.0
  Yarn: N/A
  pnpm: N/A
Relevant Packages:
  next: 15.0.3-canary.9 // Latest available version is detected (15.0.3-canary.9).
  eslint-config-next: 15.0.2
  react: 18.2.0
  react-dom: 18.2.0
  typescript: 5.6.3
Next.js Config:
  output: N/A
```
This error points to the `await headers()` function in my getIP.ts utility file.

### Which area(s) are affected? (Select all that apply)

Not sure

### Which stage(s) are affected? (Select all that apply)

next start (local)

### Additional context

I tested on nextjs 15.0.2 and 15.0.3.
On nextjs 14.2.17 works perfect.

**Comment by ztanner** [comment]
Hi, I cannot reproduce the issue as you've provided it here. The app loads with "TEST IP 1" and I don't see any errors.

Could you verify if clearing the `.next` directory inside your project helps?

**Comment by icyJoseph** [comment]
What's up with the `use server` all over the place btw?

For some reason, I kept getting this error:

```console
Error: `headers` cannot be called inside a prerender. This is a bug in Next.js.
```

Which I fixed, by adding `force-dynamic`, and then it worked. I also removed the `use server`, which seem unnecessary.

Link to Stackblitz: https://stackblitz.com/edit/github-rwpwsp?file=src%2Fapp%2Fpage.tsx,src%2Fapp%2F_util%2FgetIP.ts

Anyhow, it built correctly, no issue, and the production server runs.

```console
❯ node -v
v18.20.3
```

Just noticed that, your github repository, uses, `"next": "^15.0.3-canary.9",` though.

**Reply by MilanCiprian** [reply]
Where did you add `force-dynamic`?

**Reply by icyJoseph** [reply]
On the page itself ~ weird error though, but w.e., you can see the changes in the Stackblitz link.

**Comment by cryeprecision** [comment]
I'm getting the same issue.

Apache `2.4.62`
Node `v18.19.0`
OS `Debian GNU/Linux 12 (bookworm)`
Kernel: `Linux 6.1.0-31-amd64 x86_64`

```
"next": "^15.2.1",
"react": "^19.0.0",
"react-dom": "^19.0.0",
"eslint-config-next": "^15.2.1",
"typescript": "^5.8.2",
```

I also use `output: "standalone"` and `basePath` in my `next.config.js`.

I tried to reproduce it using docker containers (`node:18.19.0-bookworm` and `httpd:2.4.62-bookworm`) but it worked with docker.

[> Could you verify if clearing the `.next` directory inside your project helps?](https://github.com/vercel/next.js/discussions/72466#discussioncomment-11179415)
I tried that and the issue was the same.
